pennylane/ops/functions/matrix.py

```diff
@@ -211,8 +211,10 @@ def circuit():
     if isinstance(op, qml.operation.Tensor) and wire_order is not None:
         op = 1.0 * op  # convert to a Hamiltonian
 
-    if isinstance(op, qml.ops.Hamiltonian):
+    if (pr := op.pauli_rep) is not None:
+        return pr.to_mat(wire_order=wire_order or op.wires, format="csr").toarray()

```

Comment:
For context, this came up in switching to new opmath because some generators and qchem operators where now slower in computing `qml.matrix()`, as they were no longer `ops.Hamiltonian` instances and therefore didnt use the `op.sparse_matrix(..).toarray()` method below anymore. The idea was to just allow still doing that via the pauli rep in this PR; but as @albi3ro pointed out this breaks differentiability. Afaik christina has been cooking up an alternate solution 👩‍🍳